File: python/tn/handlers/outbox.py
"""Durable outbox + background worker for async handlers.

Uses persist-queue's SQLiteAckQueue: crash-safe, thread-safe, acknowledges
only after successful publish. On crash mid-publish, the item reappears at
next start.
"""

from __future__ import annotations

import logging
import random
import threading
import time
from collections.abc import Callable
from pathlib import Path
from typing import Any

try:
    from persistqueue import SQLiteAckQueue
    from persistqueue.exceptions import Empty as PQEmpty
except ImportError as e:
    raise ImportError(
        "persist-queue is required for async handlers. "
        "Install via `pip install tn-protocol` (the base install includes it)."
    ) from e

_log = logging.getLogger("tn.handlers.outbox")


class DurableOutbox:
    """Thin typed wrapper around SQLiteAckQueue."""

    def __init__(self, path: str | Path):
        self._path = Path(path)
        self._path.parent.mkdir(parents=True, exist_ok=True)
        # SQLiteAckQueue takes a directory path; creates `data.db` inside.
        self._q = SQLiteAckQueue(
            path=str(self._path),
            auto_commit=True,
            multithreading=True,
        )

    def put(self, item: dict[str, Any]) -> None:
        self._q.put(item)

    def get(self, block: bool = True, timeout: float = 1.0) -> Any:
        """Returns the raw queue item (which we later ack/nack). Raises
        persistqueue.Empty on timeout."""
        return self._q.get(block=block, timeout=timeout)

    def ack(self, item: Any) -> None:
        self._q.ack(item)

    def nack(self, item: Any) -> None:
        self._q.nack(item)

    def size(self) -> int:
        return self._q.size

    def close(self) -> None:
        self._q.close()


class OutboxWorker:
    """Background thread that drains the outbox with retries."""

    def __init__(
        self,
        outbox: DurableOutbox,
        publish: Callable[[dict[str, Any], bytes], None],
        *,
        name: str,
        max_retries: int = 10,
        backoff_initial: float = 1.0,
        backoff_max: float = 60.0,
    ):
        self._outbox = outbox
        self._publish = publish
        self._stop_ev = threading.Event()
        self._thread = threading.Thread(
            target=self._run,
            name=f"tn-outbox-{name}",
            daemon=True,
        )
        self._max_retries = max_retries
        self._backoff_init = backoff_initial
        self._backoff_max = backoff_max
        self._name = name

    def start(self) -> None:
        self._thread.start()

    def _run(self) -> None:
        while not self._stop_ev.is_set():
            try:
                item = self._outbox.get(block=True, timeout=0.5)
            except PQEmpty:
                continue
            except Exception:  # noqa: BLE001 — preserve broad swallow; see body of handler
                _log.exception("[%s] outbox.get failed", self._name)
                time.sleep(self._backoff_init)
                continue
            # Note: persist-queue passes the stored payload straight back.
            # We wrapped it as {"envelope":..., "raw":...} in AsyncHandler.emit.
            self._deliver(item)

    def _deliver(self, item: Any) -> None:
        payload = item
        envelope = payload["envelope"]
        raw = payload["raw"]

        attempt = 0
        while not self._stop_ev.is_set():
            try:
                self._publish(envelope, raw)
                self._outbox.ack(item)
                return
            except Exception as e:  # noqa: BLE001 — preserve broad swallow; see body of handler
                attempt += 1
                if attempt >= self._max_retries:
                    _log.error(
                        "[%s] publish failed after %d attempts, giving up: %s",
                        self._name,
                        attempt,
                        e,
                    )
                    # Leave nacked — it'll come back when the queue next iterates,
                    # but we want to avoid tight-looping forever. A future
                    # improvement: move to a dead-letter queue after max_retries.
                    self._outbox.nack(item)
                    return
                delay = min(
                    self._backoff_max,
                    self._backoff_init * (2 ** (attempt - 1)),
                )
                # jitter ±20% to avoid thundering herd on shared broker hiccups
                delay *= 1.0 + random.uniform(-0.2, 0.2)
                _log.warning(
                    "[%s] publish attempt %d failed: %s — retry in %.1fs",
                    self._name,
                    attempt,
                    e,
                    delay,
                )
                if self._stop_ev.wait(delay):
                    # stop requested while backing off
                    self._outbox.nack(item)
                    return

    def stop(self, *, timeout: float = 30.0) -> None:
        """Let the worker drain the outbox, then signal stop and join.

        The worker checks _stop_ev between iterations, so if we set it
        first the in-flight item completes but queued items behind it
        are left for next process start. That's correct for crashes;
        for a clean shutdown we want to wait for the outbox to empty.
        """
        deadline = time.time() + timeout
        while time.time() < deadline:
            if self._outbox.size() == 0:
                break
            time.sleep(0.1)
        self._stop_ev.set()
        remaining = max(0.5, deadline - time.time())
        self._thread.join(timeout=remaining)

File: ts-sdk/src/handlers/outbox.ts
// Durable outbox + background worker for async (network) handlers.
// Port of python/tn/handlers/outbox.py.
//
// Python uses persist-queue's SQLiteAckQueue. To keep the TS SDK
// dependency-light (no native sqlite binding), this uses an equivalent
// FILE-DIRECTORY queue with the same guarantees:
//
//   * Crash-safe: each item is its own file; `put` writes a temp file then
//     atomically renames it in. An item is `ack`ed only AFTER a successful
//     publish (by deleting its file). A crash mid-publish leaves the file on
//     disk so it's reprocessed on next start.
//   * FIFO: items are claimed oldest-first by filename (monotonic prefix).
//   * Single-process: one worker drains one item at a time on the event
//     loop; claims use an atomic rename so a half-processed item is
//     recoverable. (Multi-process draining of one dir is not a goal —
//     matches the Python single-worker model.)

import {
  existsSync,
  mkdirSync,
  readdirSync,
  readFileSync,
  renameSync,
  rmSync,
  writeFileSync,
} from "node:fs";
import { join } from "node:path";
import { randomBytes } from "node:crypto";

/** One queued item: the envelope + its raw ndjson line, as enqueued by emit. */
export interface OutboxItem {
  envelope: Record<string, unknown>;
  raw: string;
}

interface ClaimedItem {
  /** The `.processing` filename (the claim handle). */
  claim: string;
  item: OutboxItem;
}

const ITEM_SUFFIX = ".item.json";
const PROC_SUFFIX = ".processing.json";

/**
 * File-directory durable queue. Mirrors python DurableOutbox's contract:
 * `put` / `claimNext` / `ack` / `nack` / `size`, ack-after-publish, and
 * crash recovery of in-flight (`.processing`) items on construction.
 */
export class DurableOutbox {
  private readonly _dir: string;
  private _seq = 0;

  constructor(dir: string) {
    this._dir = dir;
    mkdirSync(dir, { recursive: true });
    // Crash recovery: any item left `.processing` from a previous run (the
    // process died mid-publish) is reset to pending so it's retried.
    for (const name of readdirSync(dir)) {
      if (name.endsWith(PROC_SUFFIX)) {
        const pending = name.slice(0, -PROC_SUFFIX.length) + ITEM_SUFFIX;
        try {
          renameSync(join(dir, name), join(dir, pending));
        } catch {
          /* best-effort recovery */
        }
      }
    }
  }

  /** Enqueue an item. Atomic: write temp, rename into place. */
  put(item: OutboxItem): void {
    // Monotonic-ish, sortable prefix: time + per-instance counter + random.
    const seq = (this._seq++).toString().padStart(9, "0");
    const stamp = `${Date.now().toString().padStart(15, "0")}_${seq}_${randomBytes(4).toString("hex")}`;
    const finalPath = join(this._dir, stamp + ITEM_SUFFIX);
    const tmpPath = finalPath + ".tmp";
    writeFileSync(tmpPath, JSON.stringify(item), "utf8");
    renameSync(tmpPath, finalPath);
  }

  /** Claim the oldest pending item (atomic rename to `.processing`), or
   *  null if the queue is empty. */
  claimNext(): ClaimedItem | null {
    const pending = readdirSync(this._dir)
      .filter((n) => n.endsWith(ITEM_SUFFIX))
      .sort();
    for (const name of pending) {
      const claim = name.slice(0, -ITEM_SUFFIX.length) + PROC_SUFFIX;
      try {
        renameSync(join(this._dir, name), join(this._dir, claim));
      } catch {
        // Lost the race / file vanished — try the next one.
        continue;
      }
      try {
        const item = JSON.parse(readFileSync(join(this._dir, claim), "utf8")) as OutboxItem;
        return { claim, item };
      } catch {
        // Corrupt item — drop it so it doesn't wedge the queue.
        try {
          rmSync(join(this._dir, claim));
        } catch {
          /* best-effort */
        }
        continue;
      }
    }
    return null;
  }

  /** Acknowledge (delete) a claimed item after a successful publish. */
  ack(claim: string): void {
    try {
      rmSync(join(this._dir, claim));
    } catch {
      /* already gone */
    }
  }

  /** Return a claimed item to the pending pool for retry. */
  nack(claim: string): void {
    const pending = claim.slice(0, -PROC_SUFFIX.length) + ITEM_SUFFIX;
    try {
      renameSync(join(this._dir, claim), join(this._dir, pending));
    } catch {
      /* best-effort */
    }
  }

  /** Count of pending + in-flight items. */
  size(): number {
    if (!existsSync(this._dir)) return 0;
    return readdirSync(this._dir).filter(
      (n) => n.endsWith(ITEM_SUFFIX) || n.endsWith(PROC_SUFFIX),
    ).length;
  }
}

export type PublishFn = (envelope: Record<string, unknown>, raw: string) => Promise<void> | void;

export interface OutboxWorkerOptions {
  name: string;
  maxRetries?: number;
  backoffInitialMs?: number;
  backoffMaxMs?: number;
  /** Idle poll interval when the queue is empty. */
  pollMs?: number;
  /** Deterministic jitter for tests (default: crypto-random in ±20%). */
  jitter?: () => number;
}

/**
 * Background drainer. Mirrors python OutboxWorker: claims items oldest-first,
 * retries `publish` with exponential backoff + jitter, acks on success and
 * nacks (gives up, leaving the item to retry next iteration) after
 * `maxRetries`. Runs on the event loop (no threads).
 */
export class OutboxWorker {
  private readonly _outbox: DurableOutbox;
  private readonly _publish: PublishFn;
  private readonly _name: string;
  private readonly _maxRetries: number;
  private readonly _backoffInit: number;
  private readonly _backoffMax: number;
  private readonly _pollMs: number;
  private readonly _jitter: () => number;
  private _stopped = false;
  private _runPromise: Promise<void> | null = null;

  constructor(outbox: DurableOutbox, publish: PublishFn, opts: OutboxWorkerOptions) {
    this._outbox = outbox;
    this._publish = publish;
    this._name = opts.name;
    this._maxRetries = opts.maxRetries ?? 10;
    this._backoffInit = opts.backoffInitialMs ?? 1000;
    this._backoffMax = opts.backoffMaxMs ?? 60_000;
    this._pollMs = opts.pollMs ?? 250;
    // ±20% jitter to avoid thundering-herd on a shared-broker hiccup.
    this._jitter = opts.jitter ?? (() => 1.0 + ((randomBytes(2).readUInt16BE(0) / 65535) * 0.4 - 0.2));
  }

  start(): void {
    if (this._runPromise) return;
    this._runPromise = this._run();
  }

  private async _run(): Promise<void> {
    while (!this._stopped) {
      const claimed = this._outbox.claimNext();
      if (!claimed) {
        await _sleep(this._pollMs);
        continue;
      }
      await this._deliver(claimed);
    }
  }

  private async _deliver(claimed: ClaimedItem): Promise<void> {
    let attempt = 0;
    while (!this._stopped) {
      try {
        await this._publish(claimed.item.envelope, claimed.item.raw);
        this._outbox.ack(claimed.claim);
        return;
      } catch {
        attempt += 1;
        if (attempt >= this._maxRetries) {
          // Give up on this pass; nack so it returns to pending (a future
          // dead-letter queue is a tracked improvement, matching Python).
          this._outbox.nack(claimed.claim);
          return;
        }
        let delay = Math.min(this._backoffMax, this._backoffInit * 2 ** (attempt - 1));
        delay *= this._jitter();
        const interrupted = await _sleepUntilStop(delay, () => this._stopped);
        if (interrupted) {
          this._outbox.nack(claimed.claim);
          return;
        }
      }
    }
    // Stopped mid-retry: return the item to pending.
    this._outbox.nack(claimed.claim);
  }

  /** Drain (best-effort, up to timeoutMs), then stop the run loop. */
  async stop(opts: { timeoutMs?: number } = {}): Promise<void> {
    const timeoutMs = opts.timeoutMs ?? 30_000;
    const deadline = Date.now() + timeoutMs;
    while (Date.now() < deadline && this._outbox.size() > 0) {
      await _sleep(100);
    }
    this._stopped = true;
    if (this._runPromise) {
      await Promise.race([this._runPromise, _sleep(Math.max(500, deadline - Date.now()))]);
    }
  }
}

function _sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/** Sleep that resolves early (returns true) if stop is requested. */
async function _sleepUntilStop(ms: number, stopped: () => boolean): Promise<boolean> {
  const step = 100;
  let waited = 0;
  while (waited < ms) {
    if (stopped()) return true;
    await _sleep(Math.min(step, ms - waited));
    waited += step;
  }
  return stopped();
}
